src/context.rs

```diff
@@ -76,6 +77,120 @@ impl Context {
             0
         }
     }
+
+    /// Return the stock string for the [dc_stock::StringId].
+    ///
+    /// If the context callback responds with a string to use, e.g. a
+    /// translation, then this string will be returned.  Otherwise a
+    /// default (English) string is returned.
+    pub fn stock_str(&self, id: dc_stock::StringId) -> String {
+        let ptr = self.call_cb(Event::GET_STRING, id as usize, 0) as *mut libc::c_char;
+        if ptr.is_null() {
+            dc_stock::default_string(id)
+        } else {
+            let ret = to_string(ptr);
+            unsafe { free(ptr as *mut libc::c_void) };
+            ret
+        }
+    }
+
+    /// Return stock string, replacing placeholders with provided string.
+    ///
+    /// This replaces both the *first* `%1$s` **and** `%1$d`
+    /// placeholders with the provided string.
+    pub fn stock_str_repl_string<T: AsRef<str>>(
+        &self,
+        id: dc_stock::StringId,
+        insert: T,
+    ) -> String {
+        self.stock_str(id)
+            .replacen("%1$s", insert.as_ref(), 1)
+            .replacen("%1$d", insert.as_ref(), 1)
+    }
+
+    /// Return stock string, replacing placeholders with provided int.
+    ///
+    /// Like [stock_str_repl_string] but substitute the placeholders
+    /// with an integer.
+    pub fn stock_str_repl_int(&self, id: dc_stock::StringId, insert: i32) -> String {
+        self.stock_str_repl_string(id, format!("{}", insert).as_str())
+    }
+
+    /// Return stock string, replacing 2 placeholders with provided string.
+    ///
+    /// This replaces both the *first* `%1$s` **and** `%1$d`
+    /// placeholders with the string in `insert` and does the same for
+    /// `%2$s` and `%2$d` for `insert2`.
+    pub fn stock_str_repl_string2<T: AsRef<str>>(
+        &self,
+        id: dc_stock::StringId,
+        insert: T,
+        insert2: T,
```

Comment:
this should be `S`, with the bound `AsRef<str>`, otherwise both `insert` and `insert2` have to always be the same type

Reply:
used `impl AsRef<str>` instead which i assume achieves the same